### PR TITLE
Block options: disable mode toggle when block is in warning state

### LINKED ISSUE #7743: Block option 'edit visually' and 'edit HTML' still available when block is in warning state
If you put a block into a warning state (edit HTML and remove some tags) then the block option 'Edit visually' and 'Edit as HTML' are still available and seem to toggle state, but have no apparent effect on the block.

For simplicity they should probably be removed while the block is showing a warning.

**To Reproduce**
1. Type some text in a block
2. Edit the block as HTML
3. Remove the trailing `</p>` tag
4. Click outside the block to show a warning
5. Click the blocks options to see that 'Edit visually' is still available and toggles to 'Edit as HTML' if selected, but with no change in the block

**Expected behavior**
The 'Edit visually' and 'Edit as HTML' to be disabled or not present in the menu. I think it will cause confusion when they don't do anything.

**Screenshots**
Block in warning state shows edit option:

block

Clicking this toggles the option but has no outward effect:

blok

### PR DESCRIPTION
**Pull request by johngodley**
## Description
If you put a block into a warning state the block option 'Edit visually' and 'Edit as HTML' are still available and toggle state, but have no outward effect on the block.

To avoid user confusion this PR disables the mode toggle option when a block is showing a warning. As soon as the warning is cleared the option is restored.

another_post___ _wordpress_latest_ _wordpress

Fixes #7743 

## How has this been tested?

Additional unit tests have been added to test the mode toggle enable/disable.

Manually test that the 'edit visually' and 'edit as HTML' options are removed from the block 'more options' menu:
- Add a paragraph block, edit as HTML, and remove the trailing `</p>` to trigger an invalid block warning
- Access the block 'more options' menu and verify that the 'edit visually' and 'edit as HTML' options are removed (see above screenshot)
- From a developer console type `wp.blocks.registerBlockType( 'myplugin/mr-crashy', { title: 'Mr. Crashy', category: 'common', icon: 'warning', edit() { throw new Error(); }, save() { return ''; } } );`
- Add `Mr Crashy` block to a post
- Verify that the crashed block 'more options' also does not show 'edit visually' and 'edit as HTML'

## Types of changes

This PR changes the behaviour of an existing feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by SuperGeniusZeb** [comment]
Would it make more sense to still show the options, but have them be disabled?

**Comment by aduth** [comment]
>Note this just adds the capability for additional actions, but doesn’t actually include any

[YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

File: edit-post/assets/stylesheets/_z-index.scss [hunk 1]
```diff
@@ -21,6 +20,7 @@ $z-layers: (
 	'.components-popover__close': 5,
 	'.editor-block-list__insertion-point': 5,
 	'.editor-inserter-with-shortcuts': 5,
+	'.editor-warning': 6,
```
Comment (aduth) [inline edit-post/assets/stylesheets/_z-index.scss]: Are we increasing the `z-index` for the block warning specifically, and if so, will it have negative consequences on any other "EditorWarnings" present in the application? It was meant to be made separate for general reusability, and thus we should be aware of cascading effects on other usage. Alternatively, we could create a separate `z-index` for the specific block warning.

File: editor/components/block-list/block.js [hunk 2]
```diff
@@ -558,8 +559,8 @@ export class BlockListBlock extends Component {
 							uid={ uid }
 						/>
 					) }
+					{ !! error && <BlockCrashWarning /> }
```
Comment (aduth) [inline editor/components/block-list/block.js]: Why was this moved?

**Review by aduth — COMMENTED** [review]
Aside: This pull request includes a few too many unrelated changes that's made it more difficult than it should be to review in tandem.

**Comment by johngodley** [comment]
@aduth this PR is a work in progress and contains the contents of other PRs, specifically #7866 which is handling the changes you commented on. Once the other work is finished I'll remove it from this one so it just contains the relevant code.

@SuperGeniusZeb possibly, I'll need to consult with a designer

**Comment by johngodley** [comment]
Rebased all the cruft out and went with the suggestion from @SuperGeniusZeb to disable the item, as shown in the updated screenshot. It should be good to go now.

There's a Travis CI e2e failure which seems unrelated to this PR

**Comment by johngodley** [comment]
Closing in favour of #9088
